### ISSUE TITLE
IDTReeS bugs

### ISSUE THREAD

**Issue by ashnair1** [microsoft/torchgeo#684]
### Description

While using IDTReeS to train an object detector, I noticed quite a few bugs within this dataset.

### 1. Coordinates are swapped
           
Fixed by #683

![OSBS_27_before](https://user-images.githubusercontent.com/32295036/180204371-2a1b19b8-9f47-421f-a5b9-9c82a2367f64.png)

The above image is a comparison of the actual ground truth of **OSBS_27** and the plot produced by torchgeo.

### 2. Missing polygon 

Shapefile contains 5 polygons whereas the plot only shows 4. This issue seems to be at the dataset level.

![OSBS_27_missingpoly](https://user-images.githubusercontent.com/32295036/180206989-3de7b45d-ddac-4cb1-abc0-505c9152b971.png)

### 3. Negative box coordinates in pixel space 

This is a weird one. I noticed that some boxes had negative pixel coordinates. Observed this in **MLBS_5** but its possible this could be present in others as well.

Comparing QGIS and the *corrected* torchgeo plot for **MLBS_5** didn't reveal much

![MLBS_5_initial](https://user-images.githubusercontent.com/32295036/180208170-17ef0487-a45b-46ef-8e3d-5f04acb1fcb1.png)

As you can see, the top 4 polygons are missing. But unlike the previous issue, these polygons do exist. However they seem to be outside the bounds of the image.

![MLBS_5_second](https://user-images.githubusercontent.com/32295036/180208315-718be8df-9182-4237-b9d7-7b4c7f35c7b9.png)

This could explain the negative coordinates but I'm not sure as to why there is a difference between the plots.

### Steps to reproduce

Most of these issues can be seen by downloading the IDTReeS dataset and comparing the torchgeo plots vs the shapefiles.

### Version

0.4.0.dev0

**Comment by adamjstewart** [comment]
@ashnair1 @isaaccorley any thoughts on 2 and 3? I would like to fix these issues in the 0.3.1 (coming in approximately 1 week).

Could 3 explain 2? In example 2, the missing patch seems to be right on the edge of the image.

**Comment by isaaccorley** [comment]
I think fixing the coordinates would likely fix the issue. I would bet that our bbox plotting method doesn't support boxes that are out of image bounds

**Comment by adamjstewart** [comment]
May also be related to #679, we use `int()` instead of `round()` after converting geometries. Let me fix that and we can see if it fixes this too.

**Comment by ashnair1** [comment]
Point 2 (Missing polygons) is caused by some boxes not having a classification. The issue is at the dataset level and not related to Point 3.

**Comment by ashnair1** [comment]
Point 2 (Missing polygons) is caused by the absence of some boxes in `train_data.csv`. 

The missing box here has ID `OSBS00011` which is not present in `train_data.csv`.   `itc_rsFile.csv` contains 1180 records while `train_data.csv` only contains 1166 so I'm guessing 14 boxes won't be loaded.

The issue is at the dataset level and not related to Point 3.

**Comment by ashnair1** [comment]
Point 2 (Missing polygons) is caused by the absence of some boxes in `train_data.csv`. 

The missing box here has ID `OSBS00011` which is not present in `train_data.csv`.   `itc_rsFile.csv` contains 1180 records while `train_data.csv` only contains 1166

**Comment by adamjstewart** [comment]
@ethanwhite do you know anything about points 2 and 3?

2. It seems that not all polygons have a class label
3. It seems that some polygons have coordinates outside of the bounds of the image

Are these expected? For 2, should we just ignore those polygons and pretend that they don't exist? For 3, I'm not sure how a model would be able to predict those full polygons. We could either ignore them or crop them to the bounds of the image.

**Comment by isaaccorley** [comment]
I wrote the dataset so I can provide some insight.

2. Task1 of the dataset is simply to predict the bbox of the tree crown. Task 2 is to perform both (predict + classify the species). For task1 we shouldn't worry about filtering the boxes without labels. For task2 we should filter them.

3. I think cropping to the image bounds would be better because technically it is still a tree crown it would just be a partial which would definitely be encountered during inference when tiling a larger ortho if the model was deployed. When writing the dataset I assumed all the geometries would be within the image bounds but looks like this was not the case.

**Comment by adamjstewart** [comment]
@isaaccorley sounds good to me, can you make a PR with those changes?

**Comment by ethanwhite** [comment]
@ adamjstewart - I've pinged the folks on my team who can best address this. Hopefully they'll chime in after the holiday and once we're all on the same page we can fix things in the core data source as well.

**Comment by isaaccorley** [comment]
Quick update:

I've clipped the boxes to the image bounds and seems that the boxes with negative pixel coordinates touch the image bounds but have an area of zero.

<img width="331" alt="Screen Shot 2022-09-05 at 3 50 56 PM" src="https://user-images.githubusercontent.com/22203655/188511462-fe731dde-3caa-4bf6-9be7-53e346a3094a.png">

I think the best option is to clip and remove any boxes with an area of zero. Let me know if this works for everyone.

**Comment by adamjstewart** [comment]
Hmm, that's quite a lot of images. Maybe wait for @ethanwhite's team to get back to us.

**Comment by isaaccorley** [comment]
It's 52 boxes. It affects 19 images. These are the images with boxes that only touch the image bounds.

```python
[
 'MLBS_25.tif',
 'MLBS_27.tif',
 'MLBS_32.tif',
 'MLBS_34.tif',
 'MLBS_35.tif',
 'MLBS_37.tif',
 'MLBS_39.tif',
 'MLBS_41.tif',
 'MLBS_43.tif',
 'MLBS_45.tif',
 'MLBS_5.tif',
 'OSBS_12.tif',
 'OSBS_15.tif',
 'OSBS_18.tif',
 'OSBS_2.tif',
 'OSBS_24.tif',
 'OSBS_29.tif',
 'OSBS_36.tif',
 'OSBS_38.tif'
]
```

**Comment by adamjstewart** [comment]
@isaaccorley does #760 only fix 3 or does it also fix 2? Did you want to submit a separate PR to fix 2 or should we do it in the same PR? No preference from me.

**Comment by isaaccorley** [comment]
The dataset already performs (2) by filtering boxes without labels. So this isn't technically a bug, it was just a matter of deciding to not refactor to load boxes that don't have labels.

#760 cleans up (3) though.

**Comment by adamjstewart** [comment]
> The dataset already performs (2) by filtering boxes without labels.

So you're saying the dataset already does:

> For task1 we shouldn't worry about filtering the boxes without labels. For task2 we should filter them.

So issue (2) only exists for task2, but that's okay, and issue (2) doesn't exist for task1 because we aren't filtering them?

**Comment by isaaccorley** [comment]
Yes. Right now as it's implemented, the train set doesn't have separate setup for task1/task2. It simply loads all boxes and labels and filters the rest. The test set does however support separate task1/task2.

**Comment by MarconiS** [comment]
Hi, sorry for the delay. I can help with understanding the data and figuring how to fix (2) and (3). 

> Point 2 (Missing polygons) is caused by the absence of some boxes in train_data.csv which means they have no classification.

Yes, I can confirm that for some individual trees we could not provide a species label. That is because the field sampling was performed in 20x20m quadrants and stems outside the area of interest were not identified from the field. This allowed for multiple crown objects to make it into the images while their stems did not make into the field sampling area. Since we were testing for both detection and classification, we kept these crown objects, but assigned them to no label. 

For OSBS_27 although, there seems to be another issue. The field data only cover the top-left quadrant (~10x10m), while in fact it should cover the entire image. I just checked with the field stem position data retrieved from NEON dataset, and it seems to confirm the presence of stems only where the boxes are. I am looking into it to assess whether there was an issue in retrieving stem coordinates for that specific plot. For now, I'd probably suggest to remove plot OSBS_27.

> So issue (2) only exists for task2, but that's okay, and issue (2) doesn't exist for task1 because we aren't filtering them?

This sounds  correct to me.

> It's 52 boxes. It affects 19 images. These are the images with boxes that only touch the image bounds.

I'll try to  look into it today and touch back. I need to dive into the function you are using but I believe it has to with the fact that in some cases the images may have been clipped at the exact border of some crowns.

**Comment by adamjstewart** [comment]
> Right now as it's implemented, the train set doesn't have separate setup for task1/task2

Why is that the case? I can imagine that if you're only goal is performing well on test task1, you would want to keep all boxes for train even if they don't have labels. Doesn't having different logic for train and test actually make the dataset more complicated?

**Comment by isaaccorley** [comment]
You are right. However when I initially wrote the dataset I assumed that there weren't a few boxes without labels and that boxes wouldn't be outside the bounds of some images. I'll make an issue to rewrite the dataset to have separate task1 and task2 for the train set as well.

**Comment by adamjstewart** [comment]
Cool, I think that can wait until 0.4 since that's an actual API change. Since you're making a new issue for that, I think we can consider this particular issue solved by #683 and #760. Thanks for all of the hard work everyone! Datasets are complicated!!
